pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,396 @@
+package shortnames
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/manifoldco/promptui"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// IsShortName returns true if the specified input is a "short name" and an
+// error that may occur when the input could not be parsed.  A "short name"
+// refers to a container image without a fully-qualified reference, and is
+// hence missing a registry (or domain).  Names including a digest are not
+// short names.
+//
+// Examples:
+//  * short names: "image:tag", "library/fedora"
+//  * not short names: "quay.io/image", "localhost/image:tag",
+//                     "server.org:5000/lib/image", "image@sha256:..."
+func IsShortName(input string) (bool, error) {
+	isShort, _, err := isShortName(input)
+	return isShort, err
+}
+
+// isShortName implements IsShortName but also returns the parsed
+// reference.Reference.
+func isShortName(input string) (bool, reference.Reference, error) {
+	ref, err := reference.Parse(input)
+	if err != nil {
+		return false, ref, errors.Wrapf(err, "cannot parse input: %q", input)
+	}
+
+	named, ok := ref.(reference.Named)
+	if !ok {
+		return true, ref, nil
+	}
+
+	registry := reference.Domain(named)
+	if strings.ContainsAny(registry, ".:") || registry == "localhost" {
+		return false, ref, nil
+	}
+
+	return true, ref, nil
+
+}
+
+// normalizeUserInput normalizes the user-specified reference.  Namely, it
+// strips off the tag or digest and stores it in the return values so that both
+// can be readded to possible resolved alias' value at a later point.
+func normalizeUserInput(ref reference.Reference) (isTagged bool, isDigested bool, normalized string, tag string, digest digest.Digest) {
+	normalized = ref.String()
+
+	named, isNamed := ref.(reference.Named)
+	if !isNamed {
+		return
+	}
+
+	// Strip off tag.
+	tagged, isT := named.(reference.NamedTagged)
+	if isT {
+		isTagged = true
+		tag = tagged.Tag()
+		normalized = reference.TrimNamed(tagged).String()
+		return
+	}
+
+	// Strip off digest.
+	digested, isD := named.(reference.Digested)
+	if isD {
+		isDigested = true
+		digest = digested.Digest()
+		normalized = reference.TrimNamed(named).String()
+		return
+	}
+
+	return
+}
+
+// Add records the specified name-value pair as a new short-name alias to the
+// user-specific aliases.conf.  It may override an existing alias for `name`.
+//
+// Note that tags and digests are stripped from the specified name and value.
```

Comment:
I see the value in telling everyone to use this sub package instead of the config directly (ideally the documentation in that package should point here, then).

As for the stripping, I’m worried about `podman alias fedora:latest = fedoraproject.org/fedora:f33` succeeding without error, and then `podman pull fedora:latest` surprisingly not pulling `:f33`. What will actually benefit from silently stripping the data?

Reply:
Changed. Also now everything's a `reference.Named` (will do the same for the other PR).